src/theory/strings/regexp_eval.cpp

```diff
@@ -0,0 +1,324 @@
+          // the string, each connected via single characters
+          const std::vector<unsigned>& vec = str.getVec();
+          NfaState* curr = s;
+          NodeManager* nm = NodeManager::currentNM();
+          for (size_t i = 0, nvec = vec.size(); i < nvec; i++)
+          {
+            std::vector<unsigned> charVec{vec[i]};
+            Node nextChar = nm->mkConst(String(charVec));
+            if (i + 1 == vec.size())
+            {
+              // the last edge is the dangling pointer of the first
+              sarrows.emplace_back(curr, nextChar);
+            }
+            else
+            {
+              NfaState* next = allocateState(scache);
+              curr->d_children[nextChar].push_back(next);
+              curr = next;
+            }
+          }
+        }
+      }
+      break;
+      case REGEXP_ALLCHAR:
+      case REGEXP_RANGE: sarrows.emplace_back(s, r); break;
+      case REGEXP_UNION:
+      {
+        // connect to all children via null, take union of arrows
+        std::vector<NfaState*>& schildren = s->d_children[Node::null()];
+        for (const Node& rc : r)
+        {
+          NfaState* rcs = constructInternal(rc, scache);
+          schildren.push_back(rcs);
+          std::vector<std::pair<NfaState*, Node>>& rcsarrows = rcs->d_arrows;
+          sarrows.insert(sarrows.end(), rcsarrows.begin(), rcsarrows.end());
+          rcsarrows.clear();
+        }
+      }
+      break;
+      case REGEXP_STAR:
+      {
+        NfaState* body = constructInternal(r[0], scache);
+        s->d_children[Node::null()].push_back(body);
+        // body loops back to this state
+        body->connectTo(s);
+        // skip moves on
+        sarrows.emplace_back(s, Node::null());
+      }
+      break;
+      default: Unreachable() << "Unknown regular expression " << r; break;
+    }
+    return s;
+  }
+  /** Connect dangling arrows of this to state s and clear */
+  void connectTo(NfaState* s)
+  {
+    for (std::pair<NfaState*, Node>& a : d_arrows)
+    {
+      a.first->d_children[a.second].push_back(s);
+    }
+    d_arrows.clear();
+  }
+  /** Allocate state, add to cache (for memory management) */
+  static NfaState* allocateState(std::vector<std::shared_ptr<NfaState>>& scache)
+  {
+    std::shared_ptr<NfaState> ret = std::make_shared<NfaState>();
+    scache.push_back(ret);
+    return ret.get();
+  }
+  /**
+   * Edges from this state. Maps constant characters, re.allchar or re.range
+   * to the list of children connected via an edge with that label.
+   */
+  std::map<Node, std::vector<NfaState*>> d_children;
+  /** Current dangling pointers */
+  std::vector<std::pair<NfaState*, Node>> d_arrows;
+};
+
+bool RegExpEval::canEvaluate(const Node& r)
+{
+  std::unordered_set<TNode> visited;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(r);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+
+    if (visited.find(cur) == visited.end())
+    {
+      visited.insert(cur);
```

Comment:
FYI: a better pattern for this would be to
```
auto [it, inserted] = visited.insert(cur);
if (inserted)
{
  ...
}
```
Avoids two lookups in visited.